Tutorial.md

```diff
@@ -0,0 +1,481 @@
+  sample1 <- sampleAll store
+  let expectedSample1 = M.fromList
+        [ (Identifier "my_app.requests" M.empty, Counter 1)
+        , (Identifier "my_app.connections" M.empty, Gauge 99)
+        ]
+  assert (sample1 == expectedSample1) $ pure ()
+
+  -- Replace (reregister) the connections gauge metric with a new one.
+  replacementConnectionsGauge <- Gauge.new
+  Gauge.set replacementConnectionsGauge 5
+  _ <- register store $
+    registerGauge Connections () (Gauge.read replacementConnectionsGauge)
+
+  sample2 <- sampleAll store
+  let expectedSample2 = M.fromList
+        [ (Identifier "my_app.requests" M.empty, Counter 1)
+        , (Identifier "my_app.connections" M.empty, Gauge 5)
+        ]
+  assert (sample2 == expectedSample2) $ pure ()
+
+  -- Use the dereistration handle to deregister the original metrics.
+  deregistrationHandle -- (2)
+
+  sample3 <- sampleAll store
+  let expectedSample3 = M.fromList
+        [ (Identifier "my_app.connections" M.empty, Gauge 5)
+        ]
+  assert (sample3 == expectedSample3) $ pure ()
+```
+
+1. Deregistration handles were present in in all previous examples,
+   but we ignored them for simplicity.
+
+1. The deregistration handle removes all metrics registered by the
+   initial call to `register`. In particular, this does not include the
+   reregistered gauge.
+
+## Using pre-defined sets of metrics
+
+Other libraries can define sets of metrics that you can register to your
+metric store. For example, the `ekg-core` library defines metrics for
+the runtime system metrics exposed by `GHC.Stats` -- see
+`registerGcMetrics`. Libraries that define metrics must also define
+their own metrics specifications, which you will need to include in your
+own metrics specification in order to use their metrics.
+
+Here is an example program which includes the `GcMetrics` metrics
+specification (used by `registerGcMetrics`) as a part of another metrics
+specification:
+
+```haskell
+data AppMetrics4 :: Symbol -> MetricType -> Type -> Type where
+  -- (1)
+  GcSubset ::
+    GcMetrics name metricType tags -> AppMetrics4 name metricType tags
+
+app4 :: IO ()
+app4 = do
+  store <- newStore @AppMetrics4
+  -- (2)
+  _ <- register (subset GcSubset store) registerGcMetrics
+  pure ()
+```
+
+1. We define a constructor, `GcSubset`, that takes any metric class from
+   `GcMetrics` and makes it a metric class of `AppMetrics4`.
+
+    Metric classes with the same type parameters (name, metric type, and
+    tag structure) are treated in the same way by all functions of
+    `ekg-core`, so it is enough for our constructor to "forward" the
+    type parameters.
+
+1. In order use `registerGcMetrics` with our metric store, we must use
+   the `subset` function to create a new reference to our metric store
+   annotated with the `GcMetrics` metrics specification that
+   `registerGcMetrics` expects.
+
+## Sampling subsets of metrics atomically
+
+Sampling metrics is _not_ atomic; however, subsets of metrics _can_ be
```

Comment:
it'd help to rewrite the first clause to explain that sampling multiple metrics individually in sequence is not guaranteed to be atomic — atomicity or a lack thereof isn't defined in the context of sampling a single primitive value, right?

Reply:
This is true, though ... I think I completely botched this sentence :sweat_smile: It doesn't really communicate the point I wanted to make, so I've rewritten it now.

Reply:
ah, sampleAll is not atomic, that makes sense (because it's multiple metrics)

Reply:
why not make `sampleAll` atomic by default (if that's possible)?

Reply:
or, well, rename it to `sampleAllNonAtomic` or something, and recommend using an atomic `sampleAll` function by default unless the user is happy to be provided a possibly-internally-inconsistent snapshot of the metrics in which they can use the non-atomic function instead

Reply:
Hmmmm, I think that, if the sampling actions we accept from the users can be arbitrary `IO` actions, it's not possible for us to guarantee the atomicity of a sample containing multiple metrics.

Even when we have an atomically-sampled group of metrics registered via `registerGroup`, we are relying on the user to give us an `IO` action that atomically produces the group's shared value. In that sense, the `registerGroup` function just allows a user to express a value that _they_ compute atomically in a format that `ekg-core` can understand (i.e. by projecting that value onto multiple metrics).

Reply:
ah i get it now. i think the docs could be improved to explain this better, just the way you did above, perhaps?

Reply:
I've updated the tutorial to include the bit about why sampling isn't atomic!